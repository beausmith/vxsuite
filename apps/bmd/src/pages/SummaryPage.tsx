import React, { useContext } from 'react'
import { Link, RouteComponentProps } from 'react-router-dom'
import styled from 'styled-components'

import Article from '../components/Article'
import Button from '../components/Button'
import ButtonBar from '../components/ButtonBar'
import NoPrint from '../components/NoPrint'
import { Text } from '../components/Typography'
import BallotContext from '../contexts/ballotContext'

const Header = styled.div`
  margin: 1rem 0;
`

const Table = styled.table`
  width: 100%;
  max-width: 66ch;
  text-align: left;
  border-bottom: 1px solid lightGrey;
`
const TableCell = styled.td`
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: 1px solid lightGrey;
`

const SummaryPage = (props: RouteComponentProps) => {
  const { contests, resetBallot, votes } = useContext(BallotContext)
  const startOver = () => {
    if (
      Object.keys(votes).length === 0 ||
      window.confirm('Clear all votes and start over?')
    ) {
      resetBallot()
    }
  }
  return (
    <React.Fragment>
      <Article>
        <Header className="prose">
          <h1>Official Ballot</h1>
          <p className="no-print">
            Please review your ballot. Confirm your votes by selecting the
            “Print Ballot” button.
          </p>
        </Header>
        <Table>
          <caption className="no-print visually-hidden">
            <p>Summary of your votes.</p>
          </caption>
          <thead className="no-print">
            <tr>
              <TableCell as="th" scope="col">
                Contest
              </TableCell>
              <TableCell as="th" scope="col">
                Vote
              </TableCell>
            </tr>
          </thead>
          <tbody>
            {contests.map(contest => {
              const candidate = contest.candidates.find(
                c => c.id === votes[contest.id]
              )
              const vote = candidate ? (
                candidate.name
              ) : (
                <Text muted>no selection</Text>
              )
              return (
                <tr key={contest.id}>
                  <TableCell as="th">{contest.title} </TableCell>
                  <TableCell>
                    {vote}{' '}
                    <NoPrint>
                      <small>
                        <Link to={`/contests/${contest.id}`}>change</Link>
                      </small>
                    </NoPrint>
                  </TableCell>
                </tr>
              )
            })}
          </tbody>
        </Table>
      </Article>
      <ButtonBar>
        <Button autoFocus onClick={window.print}>
          Print Ballot
        </Button>
        <Button onClick={props.history.goBack}>Back</Button>
        <Button onClick={startOver}>New Ballot</Button>
      </ButtonBar>
    </React.Fragment>
  )
}

export default SummaryPage
