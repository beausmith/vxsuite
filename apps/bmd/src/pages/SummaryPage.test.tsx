import { axe } from 'jest-axe'
import React from 'react'
import { Route } from 'react-router-dom'

import { render } from '../../test/testUtils'

import { fireEvent } from 'react-testing-library'
import SummaryPage from './SummaryPage'

it(`renders SummaryPage and request New Ballot`, () => {
  const resetBallot = jest.fn()
  const { container, getByText } = render(
    <Route path="/summary" component={SummaryPage} />,
    {
      resetBallot,
      route: '/summary',
    }
  )
  expect(container.firstChild).toMatchSnapshot()
  const newBallotButton = getByText('New Ballot')
  expect(resetBallot).not.toHaveBeenCalled()
  fireEvent.click(newBallotButton)
  expect(resetBallot).toHaveBeenCalled()
})

it(`renders SummaryPage with votes and request New Ballot`, () => {
  const resetBallot = jest.fn()
  window.confirm = jest.fn(() => true) // approve
  const { container, getByText } = render(
    <Route path="/summary" component={SummaryPage} />,
    {
      resetBallot,
      route: '/summary',
      votes: {
        president: 'minnieMouse',
        senator: 'johnSmith',
      },
    }
  )
  expect(container.firstChild).toMatchSnapshot()
  const newBallotButton = getByText('New Ballot')
  expect(resetBallot).not.toHaveBeenCalled()
  fireEvent.click(newBallotButton)
  expect(window.confirm).toBeCalled()
  expect(resetBallot).toHaveBeenCalled()
})

it(`renders SummaryPage with votes and cancels request New Ballot`, () => {
  const resetBallot = jest.fn()
  window.confirm = jest.fn(() => false) // cancel
  const { container, getByText } = render(
    <Route path="/summary" component={SummaryPage} />,
    {
      resetBallot,
      route: '/summary',
      votes: {
        president: 'minnieMouse',
        senator: 'johnSmith',
      },
    }
  )
  expect(container.firstChild).toMatchSnapshot()
  const newBallotButton = getByText('New Ballot')
  expect(resetBallot).not.toHaveBeenCalled()
  fireEvent.click(newBallotButton)
  expect(window.confirm).toBeCalled()
  expect(resetBallot).not.toHaveBeenCalled()
})

it(`empty SummaryPage is accessible`, async () => {
  const resetBallot = jest.fn()
  const { container } = render(
    <Route path="/summary" component={SummaryPage} />,
    {
      resetBallot,
      route: '/summary',
    }
  )
  expect(await axe(container.innerHTML)).toHaveNoViolations()
})

it(`SummaryPage with votes is accessible`, async () => {
  const resetBallot = jest.fn()
  const { container } = render(
    <Route path="/summary" component={SummaryPage} />,
    {
      resetBallot,
      route: '/summary',
      votes: {
        president: 'minnieMouse',
        senator: 'johnSmith',
      },
    }
  )
  expect(await axe(container.innerHTML)).toHaveNoViolations()
})
