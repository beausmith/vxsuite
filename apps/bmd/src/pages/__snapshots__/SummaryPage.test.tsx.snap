// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SummaryPage and request New Ballot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-align: center;
  margin: 1rem;
}

.c5 {
  color: gray;
}

.c1 {
  margin: 1rem 0;
}

.c2 {
  width: 100%;
  max-width: 66ch;
  text-align: left;
  border-bottom: 1px solid lightGrey;
}

.c3 {
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: none;
}

.c4 {
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: 1px solid lightGrey;
}

@media print {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    margin: 0;
  }
}

<main
  class="c0"
>
  <div
    class="prose c1"
  >
    <h1>
      Official Ballot
    </h1>
    <p
      class="no-print"
    >
      Please review your ballot. Confirm your votes by selecting the “Print Ballot” button.
    </p>
  </div>
  <table
    class="c2"
  >
    <caption
      class="no-print visually-hidden"
    >
      <p>
        Summary of your votes.
      </p>
    </caption>
    <thead
      class="no-print"
    >
      <tr>
        <th
          class="c3"
          scope="col"
        >
          Contest
        </th>
        <th
          class="c3"
          scope="col"
        >
          Vote
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          class="c4"
        >
          President
           
        </th>
        <td
          class="c4"
        >
          <span
            class="c5"
          >
            no selection
          </span>
           
          <small
            class="no-print"
          >
            <a
              href="/contests/president"
            >
              change
            </a>
          </small>
        </td>
      </tr>
      <tr>
        <th
          class="c4"
        >
          Senator
           
        </th>
        <td
          class="c4"
        >
          <span
            class="c5"
          >
            no selection
          </span>
           
          <small
            class="no-print"
          >
            <a
              href="/contests/senator"
            >
              change
            </a>
          </small>
        </td>
      </tr>
    </tbody>
  </table>
</main>
`;

exports[`renders SummaryPage with votes and cancels request New Ballot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-align: center;
  margin: 1rem;
}

.c1 {
  margin: 1rem 0;
}

.c2 {
  width: 100%;
  max-width: 66ch;
  text-align: left;
  border-bottom: 1px solid lightGrey;
}

.c3 {
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: none;
}

.c4 {
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: 1px solid lightGrey;
}

@media print {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    margin: 0;
  }
}

<main
  class="c0"
>
  <div
    class="prose c1"
  >
    <h1>
      Official Ballot
    </h1>
    <p
      class="no-print"
    >
      Please review your ballot. Confirm your votes by selecting the “Print Ballot” button.
    </p>
  </div>
  <table
    class="c2"
  >
    <caption
      class="no-print visually-hidden"
    >
      <p>
        Summary of your votes.
      </p>
    </caption>
    <thead
      class="no-print"
    >
      <tr>
        <th
          class="c3"
          scope="col"
        >
          Contest
        </th>
        <th
          class="c3"
          scope="col"
        >
          Vote
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          class="c4"
        >
          President
           
        </th>
        <td
          class="c4"
        >
          Minnie Mouse
           
          <small
            class="no-print"
          >
            <a
              href="/contests/president"
            >
              change
            </a>
          </small>
        </td>
      </tr>
      <tr>
        <th
          class="c4"
        >
          Senator
           
        </th>
        <td
          class="c4"
        >
          John Smith
           
          <small
            class="no-print"
          >
            <a
              href="/contests/senator"
            >
              change
            </a>
          </small>
        </td>
      </tr>
    </tbody>
  </table>
</main>
`;

exports[`renders SummaryPage with votes and request New Ballot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-align: center;
  margin: 1rem;
}

.c1 {
  margin: 1rem 0;
}

.c2 {
  width: 100%;
  max-width: 66ch;
  text-align: left;
  border-bottom: 1px solid lightGrey;
}

.c3 {
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: none;
}

.c4 {
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: 1px solid lightGrey;
}

@media print {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    margin: 0;
  }
}

<main
  class="c0"
>
  <div
    class="prose c1"
  >
    <h1>
      Official Ballot
    </h1>
    <p
      class="no-print"
    >
      Please review your ballot. Confirm your votes by selecting the “Print Ballot” button.
    </p>
  </div>
  <table
    class="c2"
  >
    <caption
      class="no-print visually-hidden"
    >
      <p>
        Summary of your votes.
      </p>
    </caption>
    <thead
      class="no-print"
    >
      <tr>
        <th
          class="c3"
          scope="col"
        >
          Contest
        </th>
        <th
          class="c3"
          scope="col"
        >
          Vote
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          class="c4"
        >
          President
           
        </th>
        <td
          class="c4"
        >
          Minnie Mouse
           
          <small
            class="no-print"
          >
            <a
              href="/contests/president"
            >
              change
            </a>
          </small>
        </td>
      </tr>
      <tr>
        <th
          class="c4"
        >
          Senator
           
        </th>
        <td
          class="c4"
        >
          John Smith
           
          <small
            class="no-print"
          >
            <a
              href="/contests/senator"
            >
              change
            </a>
          </small>
        </td>
      </tr>
    </tbody>
  </table>
</main>
`;
